lib/BasicAnalyses/CSVAliasAnalysis.cpp

```diff
@@ -0,0 +1,259 @@
+  I.setMetadata(LLVMContext::MD_noalias,
+                MDNode::concatenate(I.getMetadata(LLVMContext::MD_noalias),
+                                    MemoryAliasSet));
+}
+
+template<bool SegregateStackAccesses>
+PreservedAnalyses
+CSVAAP<SegregateStackAccesses>::run(Function &F, FunctionAnalysisManager &FAM) {
+  // Get the result of the GCBI analysis
+  auto &M = *F.getParent();
+  auto &MAM = FAM.getResult<ModuleAnalysisManagerFunctionProxy>(F).getManager();
+  GCBI = MAM.getCachedResult<GeneratedCodeBasicInfoAnalysis>(M);
+  if (!GCBI)
+    GCBI = &(FAM.getResult<GeneratedCodeBasicInfoAnalysis>(F));
+  revng_assert(GCBI != nullptr);
+
+  // Initialize the alias information for the CSVs.
+  initializeAliasInfo(M);
+
+  // Create the buckets in order to differentiate the accesses onto the stack
+  // from the non ones. Meanwhile, pointer arithmetic done with add instruction
+  // is transformed to use a getelementptr. Then, decorate the load/store
+  // operations with the new alias information.
+  if constexpr (SegregateStackAccesses) {
+    SegregateDirectStackAccesses SDSA(
+      static_cast<CSVAliasAnalysisInterface *>(this));
+    SDSA.segregateAccesses(F);
+  }
+
+  // Further decorate the IR with the alias information for the CSVs. Preserve
+  // the information on the stack accesses if previously added.
+  for (BasicBlock &BB : F)
+    for (Instruction &I : BB)
+      decorateMemoryAccesses(I);
+
+  return PreservedAnalyses::none();
+}
+
+using SDSA = SegregateDirectStackAccesses;
+
+void SDSA::segregateAccesses(Function &F) {
+  Value *SP = nullptr;
+
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : BB) {
+      Value *V = nullptr;
+
+      // First stage: differentiate accesses and add them onto their respective
+      // bucket.
+      if (auto *LI = dyn_cast<LoadInst>(&I)) {
+        V = skipCasts(LI->getPointerOperand());
+
+        // Continue going back if the value is a gep. Note that other pointer
+        // arithmetic done on SP is already taken into account by `skipCasts`.
+        if (isa<GetElementPtrInst>(V))
+          V = skipCasts(cast<GetElementPtrInst>(V)->getPointerOperand());
+
+        if (V == CSVAA->GCBI->spReg())
+          SP = LI;
+
+        // Everything that is not a direct access on the stack is put onto the
+        // bucket `NotDirectStackAccesses`. Load/store that access the CSVs will
+        // have their alias info added later as well.
+        if (V == SP)
+          DirectStackAccesses.push_back(&I);
+        else
+          NotDirectStackAccesses.push_back(&I);
+      } else if (auto *SI = dyn_cast<StoreInst>(&I)) {
+        V = skipCasts(SI->getPointerOperand());
+
+        if (V == SP) {
+          DirectStackAccesses.push_back(&I);
+        } else if (auto *GEP = dyn_cast<GetElementPtrInst>(V)) {
+          if (GEP->hasAllConstantIndices())
+            DirectStackAccesses.push_back(&I);
+        } else {
+          NotDirectStackAccesses.push_back(&I);
+        }
+      }
```

Comment:
There's no reason to handle `load` and `store` differently:
```c++
Value *Pointer = nullptr;
if (auto *Load = dyn_cast<LoadInst>(&I))
  Pointer = Load->getPointerOperand();
else if (auto *Store = dyn_cast<StoreInst>(&I))
  Pointer = Store->getPointerOperand();

if (Pointer != nullptr) {
  // ...
}
```

Reply:
As per below.